### PR TITLE
TSE Code Review: Ruby - Project 2

### PR DESCRIPTION
**Pull request by simoxsimo**
# [Ruby] Project 2 - Enumerable Methods

In this project I've done the following:
- Implementing custom methods identical to the enumerable methods
- Included Hashes into those methods

### CODE REVIEW

**Comment by shshamim63** [comment]
## Status :ballot_box_with_check: [!! Required Changes !!] :ballot_box_with_check: 
___
Hi @simoxsimo :wave: ,
Nice initiative, Project contains descriptive readme, PR contains a short summary. Some of the part of the project have been covered so far. But still a lot to cover. So please follow the instructions below to complete this project.
___
### :spades:  [Mandatory Changes] :spades: 
___
1. Comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/)). I can see a lot of commented code. I know they are used for checking purpose. But on your next milestone you will use rspec to test your methods. So these predefined cases are not needed. But if you want to keep it. You can use a separate file and make sure you don't push that one.
2. You have disabled couple of linter checks. Please resolve them.
____
##### Enumerable methods.
___
1. **[my_each](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))**
...1. returns an Enumerator if no block is given. 
2. **[my_each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))**
...1. returns an Enumerator if no block is given. 
3. **[my_select](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))**
...1. returns an Enumerator if no block is given. 
4. **[my_all?](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))**
...1. when no block or argument is given, returns true when none of the collection members are false or nil. `[1, 2, 3].my_all?` should return true.
...2. when a class is passed as an argument, returns true if at least one of the collection is a member of such class. `[1, 2, 3].my_all?(Integer)` should return true.
...3. when a Regex is passed as an argument, returns false if none of the collection matches the Regex. `%w[dog door rod blade].all?(/d/)` should return true.
...4. when a pattern other than Regex or a Class is given, returns false if none of the collection matches the pattern. `[1, 1, 1].my_all?(1)` should return true.
5. **[my_any?](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))**
...1.  when no block or argument is given, returns true if at least one of the collection is not false or nil. 
`[1, 2, 3].my_any?` should return true.
...2. when a class is passed as an argument, returns true if at least one of the collection is a member of such class. 
...3.when a Regex is passed as an argument, returns false if none of the collection matches the Regex.
...4. when a pattern other than Regex or a Class is given, returns false if none of the collection matches the pattern.
6. **[my_none?](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))**
...1. when no block or argument is given, returns true only if none of the collection members is true.
...2. when a class is passed as an argument, returns true if none of the collection is a member of such class
...3. when a Regex is passed as an argument, returns true only if none of the collection matches the Regex
...4. when a pattern other than Regex or a Class is given, returns true only if none of the collection matches the pattern.
7. **[my_map](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))**
...1. Returns an Enumerator if no block is given
8. **[my_inject](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))**
...1. combines all elements of enum by applying a binary operation, specified by a block.
...2.  when a symbol is specified, combines each element of the collection by applying the symbol as a named method
```ruby
[1, 2, 3].my_inject {|memo, num| memo + num } # should return 6
[1, 2, 3].my_inject(1) {|memo, num| memo + num } # should return 7
[1, 2, 3].my_inject(1, :+)  # should return 7
```
9. `multiply_els` is not an enumerable method, so it should be placed outside of Enumerable module.
___
#### Please fix the above mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/)

**Comment by simoxsimo** [comment]
# [Ruby] Project 2 - Enumerable Methods
- Fixed All issues (1..7) & 9
- After checking the issue presented by the code reviewer about my_inject method, there was no problem found.

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+    end
+
+    def my_select
+      me = self
+      temp = me.is_a?(Array) ? [] : {}
+      return to_enum(:my_select) unless block_given?
+
+      if me.is_a?(Array)
+        me.my_each { |x| temp << x if yield(x) }
+      else # if me.is_a?(Hash)
```
Comment (Forison) [inline main.rb:34]: * [ ] Make sure you don't have such ```# if me.is_a?(Hash)``` comments in your code. I understand perfectly what the comment sought to achieve but it more of a code form. Should you want to have a comment in your code, let it be plain English.

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+        me.my_each { |x, y| temp[x] = y if yield(x, y) }
+      end
+      temp
+    end
+
+    def my_all_block?(mee)
+      if mee.is_a?(Array)
+        mee.my_each { |x| return false unless yield(x) }
+      else # if mee.is_a?(Hash)
```
Comment (Forison) [inline main.rb:43]: * [ ] Make sure you don't have such ```# if me.is_a?(Hash)``` comments in your code. I understand perfectly what the comment sought to achieve but it more of a code form. Should you want to have a comment in your code, let it be plain English.

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+      ret
+    end
+  end
+
+  # contains "my_any?, my_none?" methods
+  module EnumerableInside2
+    def my_any_block?(mee)
+      if mee.is_a?(Array)
+        mee.my_each { |x| return true if yield(x) }
+      else # if mee.is_a?(Hash)
```
Comment (Forison) [inline main.rb:88]: * [ ] Make sure you don't have such ```# if me.is_a?(Hash)``` comments in your code. I understand perfectly what the comment sought to achieve but it more of a code form. Should you want to have a comment in your code, let it be plain English.

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+      temp
+    end
+
+    def my_map
+      me = self
+      return to_enum(:my_map) unless block_given?
+
+      if me.is_a?(Array)
+        my_map_array(me) { |x| yield(x) }
+      else # if me.is_a?(Hash)
```
Comment (Forison) [inline main.rb:241]: * [ ] Make sure you don't have such ```# if me.is_a?(Hash)``` comments in your code. I understand perfectly what the comment sought to achieve but it more of a code form. Should you want to have a comment in your code, let it be plain English.

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+
+        ret = my_any_array_parameter?(me, arg)
+      end
+      ret
+    end
+
+    def my_none_block?(mee)
+      if mee.is_a?(Array)
+        mee.my_each { |x| return false if yield(x) }
+      else # if mee.is_a?(Hash)
```
Comment (Forison) [inline main.rb:133]: * [ ] Make sure you don't have such ```# if me.is_a?(Hash)``` comments in your code. I understand perfectly what the comment sought to achieve but it more of a code form. Should you want to have a comment in your code, let it be plain English.

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+    end
+
+    def my_count(val = nil)
+      me = self
+      c = 0
+      return me.size if val.nil? && block_given? == false
+
+      c = if me.is_a?(Array)
+            block_given? ? a_c1(me, c) { |x| yield(x) } : a_c2(me, c, val)
+          else # if me.is_a?(Hash)
```
Comment (Forison) [inline main.rb:207]: * [ ] Make sure you don't have such ```# if me.is_a?(Hash)``` comments in your code. I understand perfectly what the comment sought to achieve but it more of a code form. Should you want to have a comment in your code, let it be plain English.

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+                my_inject_array_proc(me, start, operator)
+              end
+      start
+    end
+  end
+end
+
+def multiply_els(array)
+  array.my_inject(:*)
+  # array.my_inject() {|tot, x| tot*x} # is working too
```
Comment (Forison) [inline main.rb:300]: * [ ] Comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
``` # array.my_inject() {|tot, x| tot*x} # is working too```

**Review by Forison — APPROVED** [review]
## Status: Approved💚
# Hi @Mohammed🙂,
Good job on this one. Good job on implementing the changes asked in the previous review. I like the way you organized your code😇.
Please take note of the issues I raised below and make sure you do not repeat them in the next task or milestone. It was raised by the previous TSE but in his first observation.
# Kindly proceed to the next task👏👏👏
## Happy coding
## Addo Forison

**Pull request merged**